precompiles/outposts/stride/types.go

```diff
@@ -0,0 +1,116 @@
+// Copyright Tharsis Labs Ltd.(Evmos)
+// SPDX-License-Identifier:ENCL-1.0(https://github.com/evmos/evmos/blob/main/LICENSE)
+
+package stride
+
+import (
+	"encoding/json"
+	"fmt"
+	"math/big"
+	"strings"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/ethereum/go-ethereum/common"
+	cmn "github.com/evmos/evmos/v15/precompiles/common"
+)
+
+// StakeIBCPacketMetadata metadata info specific to StakeIBC (e.g. 1-click liquid staking)
+type StakeIBCPacketMetadata struct {
+	Action        string `json:"action"`
+	StrideAddress string
+}
+
+// Autopilot defines the receiver and IBC packet metadata info specific to the
+// Stride Autopilot liquid staking behavior
+type Autopilot struct {
+	Receiver string                  `json:"receiver"`
+	StakeIBC *StakeIBCPacketMetadata `json:"stakeibc,omitempty"`
+}
+
+// RawPacketMetadata is the raw packet metadata used to construct a JSON string
+type RawPacketMetadata struct {
+	Autopilot *Autopilot `json:"autopilot"`
+}
+
+// parseLiquidStakeArgs parses the arguments from the Liquid Stake method call
+//
+//nolint:unused
+func parseLiquidStakeArgs(args []interface{}) (common.Address, common.Address, *big.Int, string, error) {
+	if len(args) != 4 {
+		return common.Address{}, common.Address{}, nil, "", fmt.Errorf(cmn.ErrInvalidNumberOfArgs, 4, len(args))
+	}
+
+	sender, ok := args[0].(common.Address)
+	if !ok {
+		return common.Address{}, common.Address{}, nil, "", fmt.Errorf(cmn.ErrInvalidType, "sender", "", args[0])
+	}
+
+	token, ok := args[1].(common.Address)
+	if !ok {
+		return common.Address{}, common.Address{}, nil, "", fmt.Errorf(cmn.ErrInvalidType, "token", "", args[1])
+	}
+
+	amount, ok := args[2].(*big.Int)
+	if !ok {
+		return common.Address{}, common.Address{}, nil, "", fmt.Errorf(cmn.ErrInvalidType, "amount", "", args[2])
+	}
+
+	receiver, ok := args[3].(string)
+	if !ok {
+		return common.Address{}, common.Address{}, nil, "", fmt.Errorf(cmn.ErrInvalidType, "receiver", "", args[3])
+	}
+
+	// Check if the receiver address has stride before
+	if receiver[:6] != "stride" {
+		return common.Address{}, common.Address{}, nil, "", fmt.Errorf("receiver is not a stride address")
+	}
+
+	// Check if account is a valid bech32 address
+	_, err := AccAddressFromBech32(receiver, "stride")
+	if err != nil {
+		return common.Address{}, common.Address{}, nil, "", sdkerrors.ErrInvalidAddress.Wrapf("invalid stride bech32 address: %s", err)
+	}
+
+	return sender, token, amount, receiver, nil
+}
+
+// AccAddressFromBech32 creates an AccAddress from a Bech32 string.
+func AccAddressFromBech32(address string, bech32prefix string) (addr sdk.AccAddress, err error) {
```

Comment:
can we move this outside of the precompiles directory?

Reply:
and also add a unit test

Reply:
The original unit tests for this function were checked in the transaction tests.